Add check that gdb is on the path

xbauch opened the pull request:
No functional change only wrapped the test-cases in if-statements checking
that gdb can be called. The rest is clang-formatting.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

smowton commented on unit/memory-analyzer/gdb_api.cpp:
```diff
@@ -104,94 +104,111 @@ void gdb_api_internals_test()
 {
 #ifdef RUN_GDB_API_TESTS
-  gdb_api_internals_test();
+  int has_gdb = system("gdb --version");
```
How about `bool has_gdb = (system("gdb --version") == 0);`?

xbauch replied:
Done.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 65a30eaf7a3067a36d1deec1a3c4b34c9e28cfda).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105460761
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: b3c6c40d2d20a743429e7a45dbaf16759bec013e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105462720

peterschrammel commented on unit/memory-analyzer/gdb_api.cpp:
```diff
@@ -104,94 +104,111 @@ void gdb_api_internals_test()
 {
 #ifdef RUN_GDB_API_TESTS
-  gdb_api_internals_test();
+  bool has_gdb = system("gdb --version") == 0;
```
const

xbauch replied:
Added.

peterschrammel commented on unit/memory-analyzer/gdb_api.cpp:
```diff
@@ -104,94 +104,111 @@ void gdb_api_internals_test()
       }
+      catch(const gdb_interaction_exceptiont &e)
+      {
+        std::cerr
```
Don't we use message handlers in unit tests?

xbauch replied:
At least not consistently, I found a few other tests that call `std:cerr`. Should I modify this one?

tautschnig replied:
@peterschrammel What value would the use of message handlers add here?

peterschrammel commented on unit/memory-analyzer/gdb_api.cpp:
```diff
@@ -104,94 +104,111 @@ void gdb_api_internals_test()
 #ifdef RUN_GDB_API_TESTS
-  compile_test_file();
-
+  bool has_gdb = system("gdb --version") == 0;
```
const

xbauch replied:
Added.

tautschnig commented on unit/memory-analyzer/gdb_api.cpp:
```diff
@@ -104,94 +104,111 @@ void gdb_api_internals_test()
 {
 #ifdef RUN_GDB_API_TESTS
-  gdb_api_internals_test();
+  const bool has_gdb = system("gdb --version") == 0;
```
I would prefer if `run("gdb", {"gdb", "--version"})` (from `util/run.{h,cpp}`) were used in place of `system` (`system` is really heavy-weight as it starts a shell). But I do acknowledge that we already use `system` in this file, though I think that call should equally be replaced.

xbauch replied:
Replaced. The other use was replaced in the next commit.

tautschnig commented on unit/memory-analyzer/gdb_api.cpp:
```diff
@@ -104,94 +104,111 @@ void gdb_api_internals_test()
 #ifdef RUN_GDB_API_TESTS
-  compile_test_file();
-
+  const bool has_gdb = system("gdb --version") == 0;
```
As above - or maybe even factor out this check (and the `SECTION`) into a separate procedure.

xbauch replied:
Done.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 2af9fffa4f0e2d46e8f91b0ebe24b5b85da553f8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105533191

martin-cs commented:
OK, so today I'm going to be that bell-end that uses a simple patch to complain about something only semi-related.  Given that : why only gdb?  I believe there are various other tools that implement compatable text annd / or protocol interfaces.  Explicitly testing for gdb seems to tie us to that.

( Comment not review so you can (and I will not be even the slightest bit offended) ignore this. )

xbauch commented:
@martin-cs The context is that this particular unit-test hangs indefinitely if the system does not have `gdb` on the PATH. I guess if there are other tools with similar interface they don't cause this problem (or at least not when running `unit`).

tautschnig commented:
@xbauch clang-format asks for reformatting two lines.

martin-cs commented:
@xbauch : thanks for the explanation.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 96e96461b951db719fc68ab3d87012581d196718).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105574298

tautschnig commented on unit/memory-analyzer/gdb_api.cpp:
```diff
@@ -99,99 +94,117 @@ void gdb_api_internals_test()
+  {
+    REQUIRE(has_gdb);
+  }
+  return false;
```
Doesn't this now always return `false`, and thus effectively disable all the tests?

xbauch replied:
@tautschnig sorry about this type: should be fixed now.

petr-bauch committed:
Add check that gdb is on the path

No functional change.  Only wraps the test cases in an if-statements checking
the presence of gdb.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 01da54426279c28b68afbec73232c261f12810a9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105588250
